File: tests/bench/10_stdlib_str.go
// Benchmark 10: Stdlib Str — upper/lower, contains, replace, split
// Measures string manipulation overhead

package main

import (
	"fmt"
	"strings"
	"time"
)

func main() {
	// Part A: upper + lower case conversion loop
	nCase := 2_000_000
	start := time.Now()
	last := ""
	for i := 0; i < nCase; i++ {
		last = strings.ToUpper("hello world")
		last = strings.ToLower(last)
	}
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("[bench] str_case: %d upper+lower pairs in %d ms (last=%s)\n", nCase, elapsed, last)

	// Part B: contains + starts_with + ends_with checks
	base := "the quick brown fox jumps over the lazy dog"
	nSearch := 2_000_000
	start = time.Now()
	count := 0
	for i := 0; i < nSearch; i++ {
		if strings.Contains(base, "fox") {
			count++
		}
		if strings.HasPrefix(base, "the") {
			count++
		}
		if strings.HasSuffix(base, "dog") {
			count++
		}
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] str_search: %d x3 checks in %d ms (count=%d)\n", nSearch, elapsed, count)

	// Part C: replace
	nReplace := 1_000_000
	start = time.Now()
	last2 := ""
	for i := 0; i < nReplace; i++ {
		last2 = strings.ReplaceAll("hello world hello", "hello", "hi")
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] str_replace: %d calls in %d ms (last=%s)\n", nReplace, elapsed, last2)

	// Part D: split + rejoin
	nSplit := 1_000_000
	start = time.Now()
	last3 := ""
	for i := 0; i < nSplit; i++ {
		parts := strings.Split("a,b,c,d,e", ",")
		last3 = strings.Join(parts, "-")
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] str_split: %d split+join in %d ms (last=%s)\n", nSplit, elapsed, last3)
}

File: tests/bench/10_stdlib_str.rs
// Benchmark 10: Stdlib Str — upper/lower, contains, replace, split
// Measures string manipulation overhead

use std::time::Instant;

fn main() {
    // Part A: upper + lower case conversion loop
    let n_case = 2_000_000;
    let start = Instant::now();
    let mut last = String::new();
    for _ in 0..n_case {
        last = "hello world".to_uppercase();
        last = last.to_lowercase();
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] str_case: {} upper+lower pairs in {} ms (last={})", n_case, elapsed, last);

    // Part B: contains + starts_with + ends_with checks
    let base = "the quick brown fox jumps over the lazy dog";
    let n_search = 2_000_000;
    let start = Instant::now();
    let mut count: i64 = 0;
    for _ in 0..n_search {
        if base.contains("fox") { count += 1; }
        if base.starts_with("the") { count += 1; }
        if base.ends_with("dog") { count += 1; }
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] str_search: {} x3 checks in {} ms (count={})", n_search, elapsed, count);

    // Part C: replace
    let n_replace = 1_000_000;
    let start = Instant::now();
    let mut last2 = String::new();
    for _ in 0..n_replace {
        last2 = "hello world hello".replace("hello", "hi");
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] str_replace: {} calls in {} ms (last={})", n_replace, elapsed, last2);

    // Part D: split + rejoin
    let n_split = 1_000_000;
    let start = Instant::now();
    let mut last3 = String::new();
    for _ in 0..n_split {
        let parts: Vec<&str> = "a,b,c,d,e".split(',').collect();
        last3 = parts.join("-");
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] str_split: {} split+join in {} ms (last={})", n_split, elapsed, last3);
}
